src/main/java/org/candlepin/auth/CloudRegistrationAuth.java

```diff
@@ -271,19 +278,114 @@ public String generateRegistrationToken(Principal principal, CloudRegistrationIn
      *
      * @param principal
-     *  the principal for which the token will be generated
+     *     the principal for which to generate the token
      *
-     * @param ownerKey
-     *  The key of the owner/organization for which the token will be generated
+     * @param cloudRegInfo
+     *     The registration information to validate
+     *
+     * @throws UnsupportedOperationException
+     *     if the current Candlepin configuration does not support cloud registration
+     *
+     * @throws CloudRegistrationAuthorizationException
+     *     if cloud registration is not permitted for the cloud provider or account holder specified by
+     *     the cloud registration details
+     *
+     * @throws MalformedCloudRegistrationException
+     *     if the cloud registration details are null, incomplete, or malformed
+     *
+     * @return cloud authentication results including the JWT token, owner ID, and or the anonymous
+     * consumer UUID
+     */
+    public CloudAuthenticationResultDTO generateRegistrationTokenV2(Principal principal,
+        CloudRegistrationInfo cloudRegInfo)
+        throws CloudRegistrationAuthorizationException, MalformedCloudRegistrationException {
+
+        if (!this.enabled) {
+            throw new UnsupportedOperationException(
+                "cloud registration is not enabled on this Candlepin instance");
+        }
+
+        if (principal == null) {
+            throw new IllegalArgumentException("principal is null");
+        }
+
+        if (cloudRegInfo == null) {
+            throw new IllegalArgumentException("cloudRegistrationInfo is null");
+        }
+
+        CloudAuthenticationResult result = this.cloudRegistrationAdapter
+            .resolveCloudRegistrationDataV2(cloudRegInfo);
+
+        String cloudAccountId = result.getCloudAccountId();
+        if (cloudAccountId == null || cloudAccountId.isEmpty() || cloudAccountId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider account could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudProviderShortName = result.getCloudProviderShortName();
+        if (cloudProviderShortName == null || cloudProviderShortName.isEmpty() ||
+            cloudProviderShortName.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider short name could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String productId = result.getProductId();
+        if (productId == null || productId.isEmpty() || productId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("product id could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String ownerKey = result.getOwnerKey();
+        boolean ownerExistsAndPoolExists = false;
```

Comment:
nit: you could express this variable's purpose with the phrase `ownerAndPoolExist` and save the doubling of the word 'exists'.

Reply:
Yeah, that is a much better name! Updated